plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -935,9 +944,91 @@ public Optional<ConnectorTableHandle> applyDelete(ConnectorSession session, Conn
     }
 
     @Override
-    public ConnectorTableHandle beginDelete(ConnectorSession session, ConnectorTableHandle tableHandle)
+    public ConnectorTableHandle beginDelete(ConnectorSession session, ConnectorTableHandle tableHandle, RetryMode retryMode)
+    {
+        IcebergTableHandle table = (IcebergTableHandle) tableHandle;
+        if (table.getFormatVersion() < 2) {
+            throw new TrinoException(NOT_SUPPORTED, "Iceberg table updates require at least format version 2");
+        }
+        String deleteMode = table.getStorageProperties().get(DELETE_MODE);
+        if (!"merge-on-read".equals(deleteMode)) {
+            String errorMessage = "Row level deletes are only supported for tables with '" + DELETE_MODE + "' set to 'merge-on-read'.";
+            if (deleteMode != null) {
+                errorMessage += " '" + deleteMode + "' is not supported.";
+            }
+            throw new TrinoException(NOT_SUPPORTED, errorMessage);
+        }
+        verify(transaction == null, "transaction already set");
+        transaction = catalog.loadTable(session, table.getSchemaTableName()).newTransaction();
+        return table.withRetryMode(retryMode);
+    }
+
+    @Override
+    public void finishDelete(ConnectorSession session, ConnectorTableHandle tableHandle, Collection<Slice> fragments)
+    {
+        IcebergTableHandle table = (IcebergTableHandle) tableHandle;
+        Table icebergTable = transaction.table();
+
+        List<CommitTaskData> commitTasks = fragments.stream()
+                .map(slice -> commitTaskCodec.fromJson(slice.getBytes()))
+                .collect(toImmutableList());
+
+        Schema schema = SchemaParser.fromJson(table.getTableSchemaJson());
+
+        RowDelta rowDelta = transaction.newRowDelta();
+        table.getSnapshotId().map(icebergTable::snapshot).ifPresent(s -> rowDelta.validateFromSnapshot(s.snapshotId()));
+        if (!table.getEnforcedPredicate().isAll()) {
+            rowDelta.conflictDetectionFilter(toIcebergExpression(table.getEnforcedPredicate()));
```

Comment:
> can toIcebergExpression be lossy (eg compaction)? Today it cannot, but based on other use-places, it could. Do we require it to be lossless here. or is it fine if the returned expression matches some more rows than the table.getEnforcedPredicate() did? Please document

If the expression is simplified the validations will just read more parts of the manifest tree than is actually required. It may result in a conflict failing a query where it was not actually necessary, but won't allow an update that shouldn't be.

> Does it mean we support concurrent delete/delete and insert/delete, if they operate on disjoint data sets? Does it actually work? Is it tested?
> (i would expect this not to work eg because we don't support commit retries, but wonder whether it may work other than commit collision being a problem)

In theory? I do not have a test for it, but can work on adding one

> is it useful only when isolationLevel == IsolationLevel.SERIALIZABLE (next if)?

No, it's also used by `validateDataFilesExist` which should always be called